File: go/specialized/sor.go
package specialized

// SOR specialized for float64 (DoubleField)
func SOR(omega float64, g [][]float64, numIterations int) {
	m := len(g)
	n := len(g[0])

	omegaOverFour := omega * 0.25
	oneMinusOmega := 1.0 - omega
	mm1 := m - 1
	nm1 := n - 1
	for iter := 0; iter < numIterations; iter++ {
		for i := 1; i < mm1; i++ {
			for j := 1; j < nm1; j++ {
				up := g[i-1][j]
				down := g[i+1][j]
				left := g[i][j-1]
				right := g[i][j+1]
				center := g[i][j]
				neighborSum := up + down + left + right
				g[i][j] = oneMinusOmega*center + omegaOverFour*neighborSum
			}
		}
	}
}

func TestSOR(n int, iterations int) {
	if n <= 0 {
		n = 10 // default size
	}
	if iterations <= 0 {
		iterations = 100 // default iterations
	}
	omega := 1.5
	g := make([][]float64, n)
	for i := range g {
		g[i] = make([]float64, n)
		for j := range g[i] {
			if i == 0 {
				g[i][j] = 100
			}
		}
	}

	//print g
	printMatrix(g)

	SOR(omega, g, iterations)

	printMatrix(g)

}

File: typescript/specialized/sor.ts


// Specialized Successive Over-Relaxation (SOR) for number[][]

// In-place SOR for heat distribution (2D grid), matching Rust style
export function sorHeat(
  g: number[][],
  omega: number,
  numIterations: number
): void {
  const m = g.length;
  const n = g[0].length;
  const omega_over_four = omega * 0.25;
  const one_minus_omega = 1.0 - omega;
  const mm1 = m - 1;
  const nm1 = n - 1;
  for (let iter = 0; iter < numIterations; iter++) {
    for (let i = 1; i < mm1; i++) {
      const gim1 = g[i - 1];
      const gip1 = g[i + 1];
      for (let j = 1; j < nm1; j++) {
        g[i][j] = omega_over_four * (gim1[j] + gip1[j] + g[i][j - 1] + g[i][j + 1])
          + one_minus_omega * g[i][j];
      }
    }
  }
}


function main() {
  // arg 1 = grid size n (nxn)
  const n = parseInt(process.argv[2]) || 16;
  const m = n;
  const num_iterations = 1000;
  const omega = 1.5;
  // make m x n grid, all 0
  const g: number[][] = Array.from({ length: m }, () => Array(n).fill(0));
  // Set boundary conditions
  for (let i = 0; i < m; i++) {
    g[i][0] = 0.0;         // Left edge
    g[i][n - 1] = 0.0;     // Right edge
  }
  for (let j = 0; j < n; j++) {
    g[0][j] = 100.0;       // Top edge (hot)
    g[m - 1][j] = 0.0;     // Bottom edge (cold)
  }
  sorHeat(g, omega, num_iterations);
  // Optionally print the grid
  //for (const row of g) console.log(row.map(v => v.toFixed(2)).join(' '));
}

if (require.main === module) {
  main();
}

